README.md
# KidColourMobile
This is a project for Mobile System Development
Using HTML, CSS, JavaSCript


changelog.txt
========================
Kid Arithmetic Progress
========================
week10:
13/05/2018
  . Change Audio component to footer to be available all screen, user can mute or decrease volume
    wherever they are being; before player needs to click home to change audio setting
  . Start recording videos
week9:
------
10/05/2018
  . Fix time out id not well clear when game won then it still to handle the event gameover
  . Simplify the way to create home buttons manually by reuse a function createHomeButton(levelId, container)

09/05/2018
  . Working on won stars for each level, there are 3 won stars for each level;

07/05/2018
  . Fixed new updated Chrome policy about not allow autoplay audio
  . Unlocked level,
  . Game scores
  . Next level
  . gameover sound
  . colour voice when user click on
  . move colour Baskets(from left to right and vice versa)
  . move colour balls(from left to right and vice versa)

week8:
------
02/05/2018
  . Create colour balls randomly with wrong balls to drag
  . Next level

01/05/2018
  . function and animation for Game over
  . function and animation for bingo
  . Apply one code to reuse in all level with global variable of array baskets
  . Next steps is to create colour balls for each baskets varying from level1 to level 4

30/04/2018
  . Load JSon of user setting

week7:
------
29/04/2018
  . Working with timer ui and event
  . Think about score for each level, it will be increase whenever kids drag right colour ball with timer

28/04/2018
  . Defined 11 basic colours for kids
  . Level1 colour basket and colour ball with drag and drop will reuse for level and create ball
  . Think about how many ball per game, random number? how is it while nb of basket > 1 (level 2, 3, ...)

27/04/2018
  . Testing with colour balls and baskets
  . Brainstorming about logic
    . Timing while drag ball to baskets
    . Scores
    . Basket colour will be random and number of baskets will be depend on level number, ex. level 1: 1 basket, level 2: 2 baskets
    . Balls, number of diffent colour ball will be random in to the ground to drag
    . cannot drag wrong colour

26/04/2018
  . search for play next sound instead of repeat one sounds
    . autio.onended, but it is working while audio.loop = false.
  . continue for learning colour parts
    . Create basket ui with stylesheet
    . testing ball with colour

25/04/2018
  . learn colour parts
    . Ask kid to match Colours
    . Ask kid to name Colours

23/04/2018
  . More vector graphic backgrounds for day or night
  . Working on sounds and audio control

Week6:
------
22/04/2018
  . Function to change background based on day or night with getDate()

21/04/2018
  . Working on game data

20/04/2018
  . Thinking about background Image of the game to fit smart phone or iPad

Week5:
------
10/04/2018
  . Add utilities
  . Add data


week4 & intra trimester break :
------------------------------
29/03/2018
  . Including ui.js
  . Game structure and bran storming

week3:
-----
21/03/2018
. Search some images
  .fruits https://pixabay.com/en/cherry-stem-fruit-red-ripe-fresh-575547/
  .animals https://pixabay.com/en/photos/?q=baby+animal&hp=&image_type=vector&order=popular&cat=animals&min_width=&min_height=

23/03/2018
  .nature
  .Community

. Australian curriculum for kids from Kindergarten to Grade 6 of Primary school


js/ui.js
/*
  Description: Create Table element;
  Return: table element
*/
function createTable(){
  //create element P
  var element = document.createElement("table");
  return element;
}


/*
  Description: Create tr element, table row element
  Return: Tr element
*/
function createTr(){
  var element = document.createElement("tr");
  return element;
}
/*
  Param: text string, text inside the created th
  Description: Create th element
  Return: th element
*/
function createTh(text){
  var element = document.createElement("th");
  var textNote = document.createTextNode(text);
  element.appendChild(textNote);
  return element;
}
/*
  Param: text string, text of table cell
  Description: Create td element
  Return: td element
*/
function createTd(text){
  var element = document.createElement("td");
  var textNote = document.createTextNode(text);
  element.appendChild(textNote);
  return element;
}

/*
  Params: text string, text of P;
          className string, css stylesheet class name
  Description: Create dynamic paragraph with its title text
  Return: P element
*/
function createParagraph(text, className){
  //create element P
  var element = document.createElement("p");
  // create text node for p
  var t = document.createTextNode(text);

  // Append the text to <p>
  element.appendChild(t);

  if (typeof className !== 'undefined'){
    element.classList.add(className);
  }

  return element;
}

/*
  Params: text string, text of h2
          className string, css stylesheet class name, optional
  Description: To create heading H1
  Return: H2 element
*/
function createHeading1(text, className){
  var element = document.createElement("h1");

  var t = document.createTextNode(text);
  element.appendChild(t);
  if (typeof className !== 'undefined'){
    element.classList.add(className);
  }

  return element;
}

/*
  Params: text string, text of h2
          className string, css stylesheet class name, optional
  Description: To create heading H2
  Return: H2 element
*/
function createHeading2(text, className){
  var element = document.createElement("h2");

  var t = document.createTextNode(text);
  element.appendChild(t);
  if (typeof className !== 'undefined'){
    element.classList.add(className);
  }

  return element;
}

/*
  Params: text string of h3 text;
          className string, css class name, optional
  Description: To create heading H3 html element
  Return: H3 element
*/
function createHeading3(text, className){
  var element = document.createElement("h3");

  var t = document.createTextNode(text);
  element.appendChild(t);
  if (typeof className !== 'undefined'){
    element.classList.add(className);
  }
  return element;
}


/*
  Param: className strng of css class name
  Description: To create DIV
  Return: div element
*/
function createDiv(className){
  var element = document.createElement("div");

  if (typeof className !== 'undefined'){
    element.classList.add(className);
  }

  return element;
}

/*
  Params: text string of span,
          className css style
  Description: To create SPAN
  Return: span element
*/
function createSpan(text, className){
  var element = document.createElement("span");
  var t = document.createTextNode(text);
  element.appendChild(t);

  if (typeof className !== 'undefined')
    element.classList.add(className);
  return element;
}

/*
  Params: title string of text,
          className string of stylesheet class name (optional)
  Description: To create BUTTON
  Return: Button element
*/
function createButton(title, className){
  var element = document.createElement("BUTTON");        // Create a <button> element

  var t = document.createTextNode(title);       // Create a text node
  element.appendChild(t);
  if (typeof className !== 'undefined')
    element.classList.add(className);
  return element;
}


/*
  param: className optional, set style for input
  Description: To create Textbox, input text
  Return:input element
*/
function createTextBox(className){
  var element = document.createElement("input");

  //textBox.type = "text";
  element.setAttribute("type", "text");
  element.id = id;
  if (typeof className !== 'undefined'){
    element.classList.add(className);
  }

  return element;
}

/*
  Params: imageSrc string, path of image
  Description: To create image: img
  Return: Image element
*/
function createImage(imageSrc){
  var element = document.createElement("img");
  element.src = imageSrc;

  return element;
}

  /*
    Description: Create Basket div
    Return: div container as a basket with body of basket and footer of basket
  */
  function createBasket(colourName){

    //colourName = "White";
    var basket = createDiv("myBasket");

    basket.style.borderColor = colourName;
    basket.style.backgroundColor = colourName;

    basket.bodyBasket = createDiv("bodyBasket");
    if (colourName === "White")
      basket.bodyBasket.style.backgroundColor = "Black";
    basket.footerBasket = createDiv("footerBasket");
  //  footerBasket.style.backgroundColor = colourName;
    basket.appendChild(basket.bodyBasket);
    basket.appendChild(basket.footerBasket);
    basket.onclick = function(){
      playOneSound("sounds/colours/" + colourName + ".mp3");
    }
    return basket;
  }

  /*
    Description: This function is called to create ball with colour
    Return: div element of ball
  */
  function createColourBall(colourName){
    var ball = createDiv("myBall");
    ball.style.backgroundColor = colourName;
    ball.setAttribute("draggable", "true");
    return ball;
  }
  /*
    Params: limitInterval int, duration in seconds; ex. 20: it will do countdown for 20s,
    Description: This function is used to create coundown component; it return a div component and id of interval to reset if needed
    https://www.w3schools.com/howto/tryit.asp?filename=tryhow_js_countdown;
    Return: object of (timerUi:countDown component and myTimer:timer id to use next time for clearing each timer)
  */
  function createCountdown(limitInterval){
    // Set the date we're counting down to
    var countDownDate = new Date().getTime() + limitInterval * 1000;
    var countdown = createDiv("countDown");
    countdown.id = "countDown";

    // Update the count down every 1 second
    var x = setInterval(function() {

        // Get todays date and time
        var now = new Date().getTime();

        // Find the distance between now an the count down date
        var distance = countDownDate - now;

        // Time calculations for days, hours, minutes and seconds
        var days = Math.floor(distance / (1000 * 60 * 60 * 24));
        var hours = Math.floor((distance % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60));
        var minutes = Math.floor((distance % (1000 * 60 * 60)) / (1000 * 60));
        var seconds = Math.floor((distance % (1000 * 60)) / 1000);

        // Output the result in a div element
        countdown.innerHTML = seconds;

        // If the count down is over, write some text
        if (distance < 0) {
            clearInterval(x);
            countdown.innerHTML = "0";
            gameOverTimeOut();
            // document.getElementById("demo").innerHTML = "EXPIRED";
        }
    }, 1000);

    return {timerUi:countdown, myTimer:x};
  }

  /*
    Param: elementId string, id of an HTML element
    Description: Delete content of container; just make it to be reusable
  */
  function deleteContentElement(elementId){
    var element = document.getElementById(elementId);
    element.innerHTML = "";
  }

  /*
    Description: hide one div; this will remove all classes then add className hide
  */
  function hideElement(elementId){
    var element = document.getElementById(elementId);
    element.className = element.className.replace(/\bmystyle\b/g, "");
    element.classList.add("hide");
  }

  /*
    Description: This will clear all timer to repaly, after win, or game over
  */
  function clearAllTimers(){
    clearInterval(myGameTimer);
    clearInterval(moveIntervalId);
    clearInterval(moveIntervalIdR);
  }

  /*
    Description: Clear Head and Body; and also interval id
  */
  function deleteGameHeadBody()
  {
    // clearHead
    deleteContentElement("gameHead");
    // clearBody
    deleteContentElement("gameBody");

  }

  /*
    Description: This function will be called when other event set beside timer;
  */
  function gameOver(){

    // Clear all timer ids to make sure everything clear after game is over
    clearAllTimers();
    //alert("EXPIRED");
    var gameFooter = document.getElementById("gameFooter");
    gameFooter.classList.remove("hide");
    gameFooter.classList.add("footerContainer");

    // hide nextLevel;
    hideElement("goNextLevel");

    // Clear game playing UI
    //deleteGameHeadBody();
    deleteContentElement("gameBody");

    // hide Coundown
    hideElement("countDown");

    // hide go home button
    hideElement("goHome");

    var gameOverImg = createImage("images/animations/gameover.gif");
    gameOverImg.id = "gameOver";
    gameBody.appendChild(gameOverImg);
    playOneSound("sounds/actions/gameover.wav");
  }
  /*
    Description: This function is called while game over by timeout
  */
  function gameOverTimeOut(){


    clearAllTimers();
    //alert("EXPIRED");
    var gameFooter = document.getElementById("gameFooter");
    gameFooter.classList.remove("hide");
    gameFooter.classList.add("footerContainer");

    // hide nextLevel;
    hideElement("goNextLevel");

    // delete game body content
    deleteContentElement("gameBody");

    // hide Coundown
    hideElement("countDown");

    // hide go home button
    hideElement("goHome");

    var gameOverImg = createImage("images/animations/gameover.gif");
    gameOverImg.id = "gameOver";
    gameBody.appendChild(gameOverImg);
    playOneSound("sounds/actions/timeout.mp3");
  }

  /*
    Description: This function is called when user won;
  */
  function youWon(){

    // clear all timer
    clearAllTimers();

    // update total scores at game home
    var spanHome = document.getElementById("homeScore");
    spanHome.textContent = "Score: " + totalScores();
    //homeContainer.classList.remove("hide");
    var gameFooter = document.getElementById("gameFooter");
    gameFooter.classList.remove("hide");
    gameFooter.classList.add("footerContainer");

    // show nextLevel;
    if (currentGameSetting.currentLevel < gameScores.length)
    {
      console.log("You won: " + currentGameSetting.currentLevel);
      var nextLevelBtn = document.getElementById("goNextLevel");
      nextLevelBtn.classList.remove("hide");
      nextLevelBtn.classList.add("gameButton");
    }
    //deleteGameHeadBody();
    deleteContentElement("gameBody");

    // hide Coundown
    hideElement("countDown");

    // hide go home button
    hideElement("goHome");

    // Update opacities of playing stars
    updatePlayingWinStars(currentGameSetting.currentLevel);

    // update game scores
    var spanScore = document.getElementById("spanGameScore");
    spanScore.textContent = " Scores: " + totalScores() + " ";

    var gameWon = createImage("images/animations/bingo.gif");
    gameWon.id = "gameWon";
    gameBody.appendChild(gameWon);
    playOneSound("sounds/actions/cheering.mp3");

    // update star opacity to current level
    updateWinStars(currentGameSetting.currentLevel);

    // create star for next level
    if (currentGameSetting.currentLevel < gameScores.length)
      createStars(currentGameSetting.currentLevel + 1);
  }

  /*
    Description: Create button HOME
    Return: Button element
  */
  function createButtonHome(){
    // Add button HOME
    var btnHome = createButton("Home", "gameButton");
    var imgHome = createImage("images/animations/home.png");
    imgHome.classList.add("retryBackground");
    btnHome.appendChild(imgHome);

    btnHome.onclick=function(){
      homeContainer.classList.remove("hide");
      var gameFooter = document.getElementById("gameFooter");

      clearAllTimers();

      deleteGameHeadBody();

      gameFooter.classList.remove("footerContainer");
      gameFooter.classList.add("hide");
    }
    return btnHome;
  }

  /*
    Params: levelId int; ex. 1, 2, 3, ..., 12
    Description:Create 3 stars for each level by default its opacity=0.1
  */
  function createStars(levelId){
    var buttonLevel = document.getElementById("level" + levelId);
    var checkStarExisted = document.getElementById("lvl" + levelId +"_star1");
    console.log(checkStarExisted);
    if (checkStarExisted == null){ // Create only 3 stars
      for (var i=1; i<window.NumberStars + 1; i++){
        var star = createImage("images/animations/star.png");
        star.classList.add("star");
        star.style.opacity = 0.2;
        star.id = "lvl" + levelId + "_star" + i;
        buttonLevel.appendChild(star);
      }
    }
  }

  /*
    Params: levelId int; ex. 1, 2, 3, ..., 12
    Description:this will be used in Home levels reset style opacity = 1 if win one game;
                10:1 star, 20: 2stars, >30: 3stars
  */
  function updateWinStars(levelId){
    var scoreLevel = gameScores[levelId-1];
    var nbStar = scoreLevel / 10;
    if (nbStar > window.NumberStars)
      nbStar = window.NumberStars;
    for (var i=1; i < nbStar + 1; i++){
      var star = document.getElementById("lvl" + levelId + "_star" + i);
      star.style.opacity = 1; // see clearly
    }
  }

  /*
    Param: levelId int, 1, 2,3, ..., 12
    This will use to update opacity of the 3 stars at playing level;
  */
  function updatePlayingWinStars(levelId){
    var scoreLevel = gameScores[levelId-1];
    var nbStar = scoreLevel / 10;
    if (nbStar > window.NumberStars)
      nbStar = window.NumberStars;
    for (var i=1; i < nbStar + 1; i++){
      var star = document.getElementById("playing" + levelId + "_star" + i);
      star.style.opacity = 1; // see clearly
    }
  }

  /*
    param: container div HTML element; it will create Audio element and append to this container
    Description: This function is used to start a sound randomely when user start game;
  */
  function playSound(container){
      // get random number to use as index of sounds array
      var soundIndRand = getRandomInt(0, sounds.length);

      // Create Audio with id: player to play sounds
      var mySound = createAudio("sounds/" + sounds[soundIndRand]);
      mySound.id = "player";

      // play next song while finish one song rather than repeat one song during the game plays
      mySound.onended = function(){
      var newIndRand = getRandomInt(0, sounds.length);
      mySound.src = "sounds/" + sounds[newIndRand];
      console.log("Next sound: " + mySound.src);
    }
    container.appendChild(mySound);
  }


js/sound.js
/*
  This function used to create Audio to play sound;
*/
function createAudio(path){
  console.log("Start sound: " + path);
  var element = document.createElement("audio");
  //element.setAttribute("autoplay", true);
  element.autoplay = true;
  element.src = path;
  element.type = 'audio/mp3';
  element.controls = 'controls';
  element.classList.add("mute");
  element.volume = 0.1;
  return element;
}

/*
  Param: path string, directory source of video
  Description: This function is used to create Video element
  Return: video element
*/
function createVideo(path){
  var element = document.createElement("video");

  element.src = path;
  element.type = 'video/mp4';
  element.controls = 'controls';
  element.classList.add("video");
  return element;

}

/*
  Function to play a sound
*/
function playOneSound(path){
  console.log("Play sound: " + path);
  var audio = new Audio(path);
  audio.play();
}


//https://www.dl-sounds.com/royalty-free/category/children/page/3/

window.sounds = [
  "Carolans_Dream.mp3",
  "Happy_Loop.wav",
  "Hush_little_baby.mp3",
  "Little_Boy_Music_Box.wav",
  "Rock_a_bye_Baby.mp3",
  "Turkey_Purring.mp3",
  "distortion.mp3",
  "Brahms_lullaby.mp3",
  "ABC_song.mp3",
  "Digital_Kid_Loop.wav",
  "If_you_Happy.mp3",
  "Jambalaya_Loop.wav",
  "Twinkle_Little_Star.mp3"
]


js/app.js

/*
  Current game setting
*/
var currentGameSetting = {
  currentLevel: 0,
  baskets:[],
  basketIndexes: []
};

var gameScores = [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0
];



var allGameLevels = [];
var gameScore=0;
var moveIntervalId;
var moveIntervalIdR; // Use this for move baskets and balls

var totalBalls = 0;
var correctNumber =0;
var myGameTimer = null;

var mainContainer = null;
var currentPage = null;
var homeContainer = null;
var currentDraggingObject = null;
var maxNbBaskets = 4;

/*
  Get total scores during play game
*/
var totalScores = function(){
  var total = 0;
  for (var i=0; i < gameScores.length; i++){
    total += gameScores[i];
  }
  return total;
}

/*
  This function is called whenever the level clicked
*/
function resetGameVariables(gameLevel){
  console.log("resetGameVariables(): currentLevel: " + gameLevel);
  // reset array to put new colours
  currentGameSetting.basketIndexes =[];
  totalBalls = 0;
  currentGameSetting.baskets = [];
  var nbBaskets = gameLevel;
  if (gameLevel > maxNbBaskets)
    nbBaskets = maxNbBaskets;
  for (var i=0; i< nbBaskets; i++){
      var numberOfBall = getRandomInt(2, 5);
      totalBalls += numberOfBall;
      var newIndexRand = getNewRandomInt(0, colourList.length, currentGameSetting.basketIndexes);
      currentGameSetting.basketIndexes.push(newIndexRand);
      currentGameSetting.baskets.push({
        colour : colourList[newIndexRand],
        balls: numberOfBall,
      });
  }
}

/*
  Drag and Drops functions
*/
function allowDrop(event) {
    event.preventDefault();
}

/*
  This function called from objects to be draged
  event.target : dragging object
*/
function dragBallStart(event) {
    // document.body.classList.remove("errorOccurred");
    currentDraggingObject = event.target;
    event.dataTransfer.setData("text", currentDraggingObject.id);
}

/*
  basket: the div container
  This function is called while drop a ball into a basket
*/
function createOnBoxDropFunction(basket){
  // This function gets called whenever a drop event occurs
  return function(event){

    // Error checking
    if (currentDraggingObject == null) return;

    console.log("Current ball colour: " + currentDraggingObject.myColour);
    console.log("Basket colour is: " + basket.myColour);
    // Check if the shape and the box shape are the same
    if (currentDraggingObject.myColour == basket.myColour)
    {
      // Remove X
      //basket.innerHTML = "";
      basket.bodyBasket.appendChild(currentDraggingObject);
      currentDraggingObject = null;
      correctNumber++;
      // Game won
      if (correctNumber == totalBalls)
      {
        gameScores[currentGameSetting.currentLevel-1] += 10;
        console.log(gameScores);
        unlockedHomeLevels();
        clearInterval(myGameTimer);
        youWon();
      }
    }
    else {
      currentDraggingObject = null;
      //box.appendChild(spanErr);
    }

    //Must have this for dragging
    event.preventDefault();
  }
}



/*
  Reset game body; recreate baskets and balls
*/
function resetGameUI()
{
  resetGameVariables(currentGameSetting.currentLevel);

  clearAllTimers();

  var gameFooter = document.getElementById("gameFooter");
  gameFooter.classList.remove("footerContainer");
  gameFooter.classList.add("hide");
  loadBackground();

  deleteGameHeadBody();

  currentPage = null;

  // Create Head UI game
  createGameHeadUi();
  // Create Body UI game
  createGameBodyUi();
}

/*
  This function is used to create head of game; it contains countdown,
  level and score and home or try other colour while user do not like the current colour
*/
function createGameHeadUi(){

  var headContainer = document.getElementById("gameHead");
  // Coundown timer here
  var myCountdown = createCountdown(window.myTime[currentGameSetting.currentLevel-1]);
  myGameTimer = myCountdown.myTimer;
  headContainer.appendChild(myCountdown.timerUi);

  // Show level number and its stars
  var divLevel = createDiv("levelContainer");
  var imgLevel = createImage("images/numbers/nb" + currentGameSetting.currentLevel + ".png");
  imgLevel.classList.add("levelGame");
  divLevel.appendChild(imgLevel);
  headContainer.appendChild(divLevel);

  // create starts
  var checkStarExisted = document.getElementById("playing" + currentGameSetting.currentLevel +"_star1");
  if (checkStarExisted == null){ // Check to not insert stars again
    for (var i=1; i<window.NumberStars + 1; i++){
      var playingStar = createImage("images/animations/star.png");
      playingStar.classList.add("star");
      playingStar.style.opacity = 0.2;
      playingStar.id = "playing" + currentGameSetting.currentLevel + "_star" + i;
      divLevel.appendChild(playingStar);
    }
  }

  updatePlayingWinStars(currentGameSetting.currentLevel);

  // container of scores
  var divScore = createDiv("gameScore");
  var textScore = createSpan(" Scores: " + totalScores() + " ");
  textScore.id = "spanGameScore";
  divScore.appendChild(textScore);

  headContainer.appendChild(divScore);

  var goHome = createButtonHome();
  goHome.id = "goHome";
  headContainer.appendChild(goHome);
}


/*
  This function is used to create coundown, baskets and
*/
function createGameBodyUi()
{
  var container = document.getElementById("gameBody");
  correctNumber =0;

  //*******  Create Baskets   *******
  var basketContainer = createDiv("basketContainer");
  //console.log(currentGameSetting.baskets);

  for (var i=0; i < currentGameSetting.baskets.length; i++){
    console.log("Basket colour in Game Setting: " + currentGameSetting.baskets[i].colour);
    var myBasket = createBasket(currentGameSetting.baskets[i].colour);
    myBasket.id = "basket" + i;
    myBasket.myColour = currentGameSetting.baskets[i].colour;
    myBasket.ondrop = createOnBoxDropFunction(myBasket);// dropWithinBox;
    myBasket.ondragover = allowDrop;
    basketContainer.appendChild(myBasket);

    container.appendChild(basketContainer);
  }

  //*******  Create balls *******
  var ballContainer = createDiv("containerBall");

  for (var i =0; i < currentGameSetting.baskets.length; i++)
  {
    // Create balls
    for (var j=0; j < currentGameSetting.baskets[i].balls; j++)
    {
      // get disturbing ball
      var wrongBallInd = getNewRandomInt(0, colourList.length, currentGameSetting.basketIndexes);
      var wrongBall = createColourBall(colourList[wrongBallInd]);
      wrongBall.myColour = colourList[wrongBallInd];
      wrongBall.ondragstart = dragBallStart;
      ballContainer.appendChild(wrongBall);
      //
      var thisColour = currentGameSetting.baskets[i].colour;
      var ball = createColourBall(thisColour);
      ball.myColour = thisColour;
      ball.ondragstart = dragBallStart;

      //ballContainer.appendChild(ball1);
      ballContainer.appendChild(ball);
    }
  }

  switch (currentGameSetting.currentLevel) {
    case 5:
      myMoveL(ballContainer, 100);
      break;

    case 6:
      myMoveL(ballContainer, 50);
      break;

    case 7:
      myMoveL(basketContainer, 100);
      myMoveR(ballContainer, 100);
      break;

    case 8:
      myMoveR(basketContainer, 50);
      myMoveL(ballContainer, 50);
      break;

    case 9:
      myMoveToRightAndStop(ballContainer, 60); // Move to left then disappear
      break;

    case 10:
      myMoveToRightAndStop(ballContainer, 50); // Move to left then disappear
      break;

    case 11:
      myMoveR(basketContainer, 50);
      myMoveToRightAndStop(ballContainer, 40); // Move to right then disappear
      break;

    case 12:
      myMoveR(basketContainer, 50);
      myMoveToRightAndStop(ballContainer, 30); // Move to right then disappear
      break;

    default:
      break;
  }
  container.appendChild(ballContainer);

}

/*
  This function is called to create Footer of Game; When game won, when game over this will create button retry or next
*/
function createGameFooterUi()
{
  // ******** Footer buttons, replay, next colour, next level
  var footerDiv = createDiv();
  footerDiv.id = "gameFooter";

  // Add button HOME
  var btnHome = createButtonHome();
  footerDiv.appendChild(btnHome);

  // Button Try again
  var btnTryAgain = createButton("Play again", "gameButton");
  // Set event to play with next colour
  btnTryAgain.onclick = function(){
    resetGameUI();
  };
  // image retry or refresh
  var imgRetry = createImage("images/animations/retry.png");
  imgRetry.classList.add("retryBackground");
  btnTryAgain.appendChild(imgRetry);

  footerDiv.appendChild(btnTryAgain);

  // Button go to next level
  var btnNextLevel = createButton("Next level", "gameButton");
  btnNextLevel.id = "goNextLevel";
  btnNextLevel.onclick = function(){
    if (currentGameSetting.currentLevel < allGameLevels.length){
        currentGameSetting.currentLevel++;
        resetGameUI();
    }
  }
  footerDiv.appendChild(btnNextLevel);

  var imgNext = createImage("images/animations/next1.png");
  imgNext.classList.add("retryBackground");
  btnNextLevel.appendChild(imgNext);

  return footerDiv;
}

/*
  This is the main function to load all scenses of the game
*/
function loadingGameLevel() {

    //console.log("Level: " + event.target.id);
    //var selectedLevel = parseInt(event.target.id);
    //currentGameSetting.currentLevel = selectedLevel;
    resetGameVariables(currentGameSetting.currentLevel);

    if (homeContainer != null)
      homeContainer.classList.add("hide");
    // hide any existing page
    if (currentPage != null)
      currentPage.classList.add("hide");
    //console.log(typeof window.game1Page);
    // check if we haven’t yet created this page
    if (typeof window.game1Page === "undefined") {

      // console.log("Start Game1 Page");

      window.game1Page = createDiv();
      window.game1Page.id = "pageColours";
      //game1Page.classList.add("page");
      mainContainer.appendChild(window.game1Page);

      //*******
      var gameHead = createDiv();
      gameHead.id = "gameHead";
      var gameBody = createDiv();
      gameBody.id = "gameBody";
      window.game1Page.appendChild(gameHead);
      window.game1Page.appendChild(gameBody);
    }

    // ******** create Head UI of game;
    createGameHeadUi();
    // ******** Create Body UI of Game
    createGameBodyUi();

    var footerGame = createGameFooterUi();

    game1Page.appendChild(footerGame);
    footerGame.classList.add("hide");

    // set the new currentPage to this game
    currentPage = game1Page;
    // show the page (now it is definitely created)
    currentPage.classList.remove("hide");
}



var muted = 0;
/*
  Loading image background randomely for day or night
*/
function loadBackground(){
  // random for change image
  var randomNum = getRandomInt(1, 13);
  var backgroundImage = "images/background/"+ whatTime()+ randomNum+ ".png";
  // Set background image
  document.body.style.backgroundImage = "url('" + backgroundImage + "')";//"url'"+(backgroundImage)+"'";

}

/*
  Onclick for muted and unmuted;
*/
function onMute(event){
  document.getElementById('player').muted=!document.getElementById('player').muted;
  muted++;
  var muteStatus = muted % 2;
  event.target.src = "images/animations/mute" + muteStatus + ".png"
}

/*
  Param: level int, ex. 1, 2, 3, ..., 12
  This is to reuse to insert locked image when game not yet reach the level;
  it starts from level 1, if user completed level 1 then level 2 will be unlocked
*/
function lockedLevel(level){
  // create image locked
  var lockedImg = createImage("images/animations/locked.png");
  lockedImg.classList.add("locked");
  level.appendChild(lockedImg);
  return lockedImg;
}

/*
  Create Game UI at start, this is one time used function
*/
function createGameHomeUi(){

  // create main container
  mainContainer = createDiv();
  document.body.appendChild(mainContainer);

  // This container contains Game title and menu to be hidden when playing
  homeContainer = createDiv();

  // Home, game title
  var headContainer = createDiv("title");
  var spanScore = createSpan("Score: 0");
  spanScore.id = "homeScore";
  headContainer.appendChild(spanScore);
  var h1 = createHeading1("Learning Colours for Kids");
  headContainer.appendChild(h1);



  // Create Menu of levels
  var menuContainer = createDiv("menu");

  var footerHomeContainer = createDiv();
  var kidImg = createImage("images/animations/kids.png");
  kidImg.classList.add("kids");
  footerHomeContainer.appendChild(kidImg);
  menuContainer.appendChild(footerHomeContainer);

  //menuContainer.appendChild(level1);

  var button1 = createButton("", "gameMenu");
  button1.id = "level1";
  button1.onclick = function(){
    currentGameSetting.currentLevel = 1;
    loadingGameLevel();
  };
  // image number
  var level1 = createImage("images/numbers/nb1.png");
  level1.classList.add("level");
  // Add image into menu buttons
  button1.appendChild(level1);
  allGameLevels.push(button1);
  menuContainer.appendChild(button1);

  // Create home button 2 to 12
  for (var i=2; i <= gameScores.length; i++){
    createHomeButton(i, menuContainer);
  }

  homeContainer.appendChild(headContainer);
  homeContainer.appendChild(menuContainer);
  mainContainer.appendChild(homeContainer);

  // ****** Footer of game
  var divFooter = createDiv("footer");
  mainContainer.appendChild(divFooter);

  // play sounds;
  var divSound = createDiv();
  divFooter.appendChild(divSound);
  playSound(divSound);

}

/*
  This is reusable component, it will create one menu button;
  Example: level 2, 3, 4, ...
*/
function createHomeButton(levelId, menuContainer){
  var button = createButton("", "gameMenu");
  button.id = "level"+levelId;
  var level = createImage("images/numbers/nb"+ levelId +".png");
  level.classList.add("level");
  button.appendChild(level);
  var imgLocked = lockedLevel(button);
  imgLocked.id = "imgLocked"+levelId;
  allGameLevels.push(button);
  menuContainer.appendChild(button);
}

/*
  This function will reset each level button for every game win; to remove locked and put the onclick event handler
*/
function unlockedHomeLevels(){
  // ex: current level is 1; it is going to unlocked level 2;
  if (currentGameSetting.currentLevel < gameScores.length) // do not take action for level 12
  {
    var nextLevel = currentGameSetting.currentLevel + 1;
    console.log("unlockedHomeLevels(): start unlocked levels: " + nextLevel);
    var btnNextLevel = allGameLevels[nextLevel - 1];
    console.log(btnNextLevel);
    // set onclick
    btnNextLevel.onclick = function(){
      currentGameSetting.currentLevel = nextLevel;
      loadingGameLevel();
    }

    // hide locked image
    var imgLocked = document.getElementById("imgLocked" + nextLevel);
    imgLocked.classList.remove("locked");
    imgLocked.classList.add("hide");
  }
}

/*
  Starting game; Load the user interfaces
*/
function startGame(){
  //playSound("../../sounds/distortion.mp3");
  //palySoundLoop();
  console.log("Game start!");
  //document.body.innerHTML = "";
  loadBackground();

  // Loading Home page of Game
  createGameHomeUi();

  createStars(1);// create 5 stars with opacity:0.2 at start with level1; will create stars for next level while won level 1;


}

window.onload = startGame;
